hydra-node/src/Hydra/Network/Etcd.hs

```diff
@@ -0,0 +1,513 @@
+          onConnectivity $ HandshakeFailure{remoteHost, ourVersion, theirVersions}
+
+      let value = event ^. #kv . #value
+      case decodeFull' value of
+        -- TODO: error handling (with actual client)
+        Left err -> fail $ "Failed to decode etcd entry: " <> show err
+        Right msg -> deliver msg
+
+getLastKnownRevision :: MonadIO m => FilePath -> m Natural
+getLastKnownRevision directory = do
+  liftIO $
+    try (decodeFileStrict' $ directory </> "last-known-revision.json") >>= \case
+      Right rev -> do
+        pure $ fromMaybe 1 rev
+      Left (e :: IOException)
+        | isDoesNotExistError e -> pure 1
+        | otherwise -> do
+            fail $ "Failed to load last known revision: " <> show e
+
+putLastKnownRevision :: MonadIO m => FilePath -> Natural -> m ()
+putLastKnownRevision directory rev = do
+  liftIO $ encodeFile (directory </> "last-known-revision.json") rev
+
+-- | Write a well-known key to indicate being alive, keep it alive using a lease
+-- and poll other peers enries to yield connectivity events. While doing so,
+-- overall network connectivity is determined from the ability to read/write to
+-- the cluster.
+pollConnectivity ::
+  Connection ->
+  -- | Local host
+  Host ->
+  NetworkCallback msg IO ->
+  IO ()
+pollConnectivity conn localHost NetworkCallback{onConnectivity} = do
+  seenAliveVar <- newTVarIO []
+  withGrpcContext "pollConnectivity" . forever $ do
+    leaseId <- createLease
+    -- If we can create a lease, we are connected
+    onConnectivity NetworkConnected
+    handle onGrpcException $ do
+      -- Write our alive key using lease
+      writeAlive leaseId
+      withKeepAlive leaseId $ \keepAlive ->
+        forever $ do
+          -- Keep our lease alive
+          keepAlive
+          -- Determine alive peers
+          alive <- getAlive
+          let othersAlive = alive \\ [localHost]
+          seenAlive <- atomically $ swapTVar seenAliveVar othersAlive
+          forM_ (othersAlive \\ seenAlive) $ onConnectivity . PeerConnected
+          forM_ (seenAlive \\ othersAlive) $ onConnectivity . PeerDisconnected
+          threadDelay 1
+ where
+  ttl = 3
+
+  onGrpcException GrpcException{grpcError = GrpcUnavailable} = do
+    onConnectivity NetworkDisconnected
+    threadDelay 1
+  onGrpcException e = throwIO e
+
+  -- REVIEW: server can decide ttl?
+  createLease = withGrpcContext "createLease" $ do
+    leaseResponse <-
+      nonStreaming conn (rpc @(Protobuf Lease "leaseGrant")) $
+        defMessage & #ttl .~ ttl
+    pure $ leaseResponse ^. #id
+
+  writeAlive leaseId = withGrpcContext "writeAlive" $ do
+    void . nonStreaming conn (rpc @(Protobuf KV "put")) $
+      defMessage
+        & #key .~ "alive-" <> show localHost
+        & #value .~ serialize' localHost
+        & #lease .~ leaseId
+
+  withKeepAlive leaseId action = do
+    biDiStreaming conn (rpc @(Protobuf Lease "leaseKeepAlive")) $ \send _recv ->
```

Comment:
We experimented with hiding this boilerplate in https://hackage.haskell.org/package/grapesy-etcd-0.3.0/docs/Network-GRPC-Etcd-Client-Simple.html. We could hide all of it if we want. `biDiStreaming`, `NextElem`, and the `rpc` type application could all be encapsulated for all calls like this.